torch/distributed/fsdp/sharded_grad_scaler.py

```diff
@@ -183,100 +147,49 @@ def _foreach_non_finite_check_and_unscale_cpu_(
-                    else:
-                        to_unscale = param.grad
-
-                    per_device_and_dtype_grads[to_unscale.device][
-                        to_unscale.dtype
-                    ].append(to_unscale)
-
-            for device, per_dtype_grads in per_device_and_dtype_grads.items():
-                for grads in per_dtype_grads.values():
-                    if grads[0].device.type == "cpu":
-                        self._foreach_non_finite_check_and_unscale_cpu_(
-                            grads,
-                            per_device_found_inf.get(device),
-                            per_device_inv_scale.get(device),
-                        )
-                    else:
-                        torch._amp_foreach_non_finite_check_and_unscale_(
-                            grads,
-                            per_device_found_inf.get(device),
-                            per_device_inv_scale.get(device),
-                        )
+        per_device_found_inf_dict = super()._unscale_grads_(
+            optimizer, inv_scale, found_inf, allow_fp16
+        )
         # There exist contexts (e.g. w/ `use_orig_params=True`) wherein some
         # ranks may have no (non-zero sized) parameter shards, necessitating the
         # initialization of `per_device_found_inf._per_device_tensors` here
-        if not per_device_found_inf._per_device_tensors:
+        if not per_device_found_inf_dict:
             assert self._scale is not None
+            per_device_found_inf = _MultiDeviceReplicator(found_inf)
             per_device_found_inf.get(self._scale.device)
-        return per_device_found_inf._per_device_tensors
+            return per_device_found_inf._per_device_tensors
+        else:
+            return per_device_found_inf_dict
 
     def unscale_(self, optimizer: torch.optim.Optimizer) -> None:
-        if not self._enabled:
-            return
-
-        self._check_scale_growth_tracker("unscale_")
-
-        optimizer_state = self._per_optimizer_states[id(optimizer)]
-
-        if optimizer_state["stage"] is OptState.UNSCALED:
-            raise RuntimeError(
-                "unscale_() has already been called on this optimizer since the last update()."
-            )
-        elif optimizer_state["stage"] is OptState.STEPPED:
-            raise RuntimeError("unscale_() is being called after step().")
-
-        # FP32 division can be imprecise for certain compile options, so we carry out the reciprocal in FP64.
-        assert self._scale is not None
-        inv_scale = self._scale.double().reciprocal().float()
-        found_inf = torch.full(
-            (1,), 0.0, dtype=torch.float32, device=self._scale.device
-        )
-
-        optimizer_state["found_inf_per_device"] = self._unscale_grads_(
-            optimizer, inv_scale, found_inf, True
-        )
-        optimizer_state["stage"] = OptState.UNSCALED
+        super().unscale_(optimizer)
```

Comment:
You can just delete this and it'll automatically fall back

Reply:
you mean delete `unscale_` method from ShardedGradScaler? I tried but CI warns me about BC breaking. should I just delete and ignore the BC warning?